client/build_test.go

```diff
@@ -2051,6 +2052,40 @@ func testClientGatewayFrontendAttrs(t *testing.T, sb integration.Sandbox) {
 	checkAllReleasable(t, c, sb, true)
 }
 
+func testClientGatewayNilResult(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+	c, err := New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+		diff := llb.Diff(st, st)
+		def, err := diff.Marshal(sb.Context())
+		if err != nil {
+			return nil, err
+		}
+		res, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+			Evaluate:   true,
+		})
+		require.NoError(t, err)
+
+		ref, err := res.SingleRef()
+		require.NoError(t, err)
+
+		// verify that this returns an error, but doesn't panic buildkitd
```

Comment:
Shouldn't the `ref` also be `nil` in this case. Or at least there should be some saner way to check for nil ref from the client side.

Reply:
I agree but it seems tricky to deal with the case where evaluate isn't true. You won't know the actual result is `nil` until its evaluated, but you have to return something to represent it until then. 

I guess it would be possible to only return nil only when evaluate is true, but then you still have to deal with the case where something is "lazily nil" when evaluate is false, so I don't know if that actually simplifies anything overall.